### PR TITLE
Let compiler run on more restrictive platforms like WASM

### PR DESCRIPTION
**Pull request by nike4613**
This PR does a couple of things:
- It marks as `lazy` most accesses anything in `System.Diagnostics.Process`
  - Most of these are not actually needed for embedded compilation anyway, so this doesn't cause any issues for basic usage
- It tests `Environment.OSVersion.Platform` in `GetPathToDotNetFrameworkImlpementationAssemblies ` so that it doesn't call into MSBuild utilities on the `Other` platform, as that tries to spin up some threads which access `System.Diagnostics.Process` and cause errors when running in Blazor.
- It rethrows more internal exceptions in a way that is more debuggable from the outside, preserving stack traces from deep internal errors.
- It removes `FSharpEnvironment.fsharpCompilerPath` because it was not used, and its eager evaluation caused errors in Blazor.

I am almost certain that this will need a little bit of cleanup or refactoring, but the increased portability would be incredibly nice to have.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12163) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12163) before we can accept your contribution.<br/><br/>:x: nike4613 [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12163)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=12163) it.</sub>

**Comment by dsyme** [comment]
@nike4613 A couple of things

1. Are the restrictions of "platforms like WASM" documented anywhere?
2. We'd need automated testing to make this worthwhile

Overall we'd also need to be confident that we could understand and fix failures should we hit restrictions. That's tricky without deeper knowledge of the toolchain. We'd certainly need lots of documentation and links.  Also if something goes wrong, we'd likely turn off WASM testing and support.

An alternative can be for the part of the F# community who care about running on WASM maintain a fork where you can arrange for WASM-based automated testing to happen.   That's what happens for FCS running on Fable, for example.

**Comment by nike4613** [comment]
Unfortunately, as far as I can tell, there is no single source of documentation for the restrictions of any specific platform. However, the BCL has [UnsupportedOSPlatforms](https://github.com/dotnet/runtime/search?q=UnsupportedOSPlatforms&type=code) properties, which generate [UnsupportedOSPlatform](https://github.com/dotnet/runtime/search?q=UnsupportedOSPlatform&type=code) attributes, and there are also [SupportedOSPlatform](https://github.com/dotnet/runtime/search?q=SupportedOSPlatform&type=code) attributes, all of 
 which describe which platforms different APIs are available on. This information could be used to generate a list of restricted APIs, but I don't know if one is currently hosted anywhere.

Roslyn has a [platform compatibility analyzer](https://github.com/dotnet/roslyn-analyzers/pull/3917) which emits warnings when using an API which are potentially incompatible with some platforms, even taking into account basic runtime checks to `Environment.OSVersion`. I'm sure it would be possible to have some kind of similar analyzer here, though it would need to be built up from scratch, and it would likely even be sufficient to have a general post-processing step that runs on the output assemblies to determine which parts of the API surface call into platform-specific BCL members.

My current experience working with Blazor WASM has been that, excluding APIs which are unavailable, if it works on a desktop runtime, then it will work in-browser. I would even argue that if there was a case where that was *not* true, then that would be a runtime bug in Mono. That being said, I'll probably look into automated testing of WASM assemblies.

**Comment by KevinRansom** [comment]
The PreserveStackTrace change is a good change regardless of WASM.

File: src/fsharp/fsi/fsi.fs [hunk 1]
```diff
@@ -204,16 +204,16 @@ module internal Utilities =
 [<AutoSerializable(false)>]
 type internal FsiTimeReporter(outWriter: TextWriter) =
     let stopwatch = Stopwatch()
-    let ptime = Process.GetCurrentProcess()
+    let ptime = lazy Process.GetCurrentProcess()
```
Comment (KevinRansom) [inline src/fsharp/fsi/fsi.fs:207]: Why is this changed to Lazy?  I don't really understand why it would be necessary?

Comment (nike4613) [reply src/fsharp/fsi/fsi.fs:207]: If it isn't, then that call is evaluated eagerly when `FsiTimeReporter` is constructed, even if its never actually used. I admit, it was more of a quick way to make it behave where `Process` isn't available, and it would probably be better to lazily construct `FsiTimeReporter` instead, but this was easier to find and solved `Process` throwing.

Comment (KevinRansom) [reply src/fsharp/fsi/fsi.fs:207]: I think I get it.  Is Process.GetCurrentProcess() particularly expensive?

Comment (nike4613) [reply src/fsharp/fsi/fsi.fs:207]: No, the problem is that on WASM it throws unconditionally. By making it lazy, it's never evaluated until its needed for timing, which is off by default.

**Comment by dsyme** [comment]
I'm going to close this because merging it is only possible with full automated testing on WASM, and that's too subtle for this repo, especially given the kinds of changes needed.  Like the Fable port of F# Compiler Service, it's currently best done out of repo by a compiler for with automatic merging of changes here.
